### ISSUE TITLE
Нужно ограничение на количество неудачных запросов.

### ISSUE THREAD

**Issue by IATkachenko** [IATkachenko/HA-YandexWeather#71]
```
Да, и еще (для Игоря) у меня за несколько часов интеграция съела все 50 запросов от Яндекса. 
Похоже, что когда возникает подобная ошибка, интеграция ломится в Яндекс и сжирает весь лимит 50 запросов в день.
```
_Originally posted by @alexanderznamensky in https://github.com/IATkachenko/HA-YandexWeather/issues/70#issuecomment-1492852330_

**Comment by IATkachenko** [comment]
@alexanderznamensky, А можно по подробнее? У меня в логе интеграция исправно раз в полчаса крашилась?

**Comment by alexanderznamensky** [comment]
Игорь, добрый день.

На двух серверах превышено количество запросов (это я по Яндексу определил).
![image](https://user-images.githubusercontent.com/65254375/229281352-c6297109-7057-4371-ac6e-e3175d243fca.png)

Чтобы я вам опять много лишней инфы не отправил, скажите, пож, что нужно посмотреть?

**Comment by heggi** [comment]
У меня было такое же, но это из-за того что я много раз перезагружал интеграцию и редактировал код в экспериментах. А так оно стабильно крашилось раз в пол часа.

**Comment by IATkachenko** [comment]
@alexanderznamensky, я даже не представляю как можно попробовать отладиться... по идее у вас в логах должны быть ошибки от интеграции при обновлении данных с интервалом меньшем, чем интервал обновления (30 минут, если использовать дефолтные 48 запросов в сутки).
Я попробую на стенде привнести ошибок в процесс, чтобы посмотреть, но пока я не очень понимаю как у вас это получилось. Версия одна (и ее планирую проверять) -- перезапуск ХА при отсутствии сохраненных данных и ошибках в процессе обновления (см ниже).

@heggi, у интеграции работает такой механизм: при перезапуске пытаемся взять сохраненные данные, если не получилось или они слишком старые -- идем обновлять. А дальше просыпается механизм ХА: интеграция не справилась стартануть -- попробуем еще раз. Поэтому ваше поведение вполне логичное: сохраненных данных нет, обновляемся при каждом перезапуске до посинения. Защищаться нужно, видимо, от этого...

**Comment by heggi** [comment]
Нет, HA не пробовал перезапускать интеграцию при неудачном старте. Это я сам ее перезапускал, когда пытался заставить измененный код работать (не силен в питоне, пришлось с помощью гугла и какой-то матери понимать что там происходит).
Так что если интеграцию не трогать (не перезагружать), поведение получаем вполне логичное и предсказуемое.

**Comment by IATkachenko** [comment]
Он это в фоне, обычно, делает. В логе там строчка вида: стартануть не получилось -- продолжаем пробовать в фоне.

В любом случае, спасибо за наводку и сценарии катастроф -- погоняю у себя.

**Comment by alexanderznamensky** [comment]
я думаю, что виной может являться моя автоматизация, которая раз в 30 минут ее перезапускает:
  - id: "Перезагрузка интеграции Yandex Weather"
    alias: "reload_yandex_weather"
    initial_state: true
    trigger:
      - platform: time_pattern
        minutes: '/30'
    condition:
      - condition: state
        entity_id: weather.yandex_weather
        state: 'unavailable'
    action:
# Перезагрузка интеграции
      - service: homeassistant.reload_config_entry
        data: {}
        target:
          device_id: b2ddcdca6887819edc8e6ad4fa0ccc19
    mode: single

**Comment by AlexeyNapalkov** [comment]
Хотел сегодня подключить вашу интеграцию
завел ключ выждал 30 минут

но при попытке подключения форма получаю ошибку (координаты в тексте я заменил ХХ.Х)
При этом Яндекс считает количество использования ключа для каждой моей попытки настроить интеграцию.

======== Текст ошибки из журнала HA ========
Эта ошибка возникла в пользовательской интеграции.

Logger: custom_components.yandex_weather.updater
Source: custom_components/yandex_weather/updater.py:308
Integration: Яндекс.Погода (documentation, issues)
First occurred: 13:06:57 (1 occurrences)
Last logged: 13:06:57

Could not get data from API: <ClientResponse(https://api.weather.yandex.ru/v2/informers?lat=ХХ.Х5996353910978&lon=ХХ.Х0080108642579&lang=en_US) [403 Forbidden]> <CIMultiDictProxy('Content-Encoding': 'gzip', 'Content-Length': '52', 'Content-Type': 'text/plain; charset=utf-8', 'Date': 'Wed, 25 Oct 2023 10:06:57 GMT', 'Vary': 'Accept-Encoding', 'X-Content-Type-Options': 'nosniff', 'X-Req-Id': '1698228417029102-9013365984183450632-hxachmby352st2km-BAL')>

**Comment by IATkachenko** [comment]
План такой: три подряд неудачных обновления -- следующее запускаем не раньше чем через час.

**Comment by dimastark** [comment]
Официальная интеграция с Яндекс Погодой находится в организации yandex
https://github.com/yandex/pogoda-home-assistant

В ней решены подобные проблемы с количеством запросов, она развивается (туда недавно добавили время восхода и заката), учитываются фич-реквесты.

Поэтому, всегда можно переключиться на неё.

**Comment by IATkachenko** [comment]
> Официальная интеграция с Яндекс Погодой находится тут https://github.com/yandex/pogoda-home-assistant
> 
> Она является форком этого проекта (спасибо большое [@IATkachenko](https://github.com/IATkachenko) за помощь, который также является основным контрибьютером проекта в огранизации yandex, при этом мы не против параллельного развития двух вариантов плагина).
> 
> В новой интеграции нет проблем с количеством запросов, потому что тариф "Умный дом" был специально создан для нового плагина.

Увы, если что-то пойдет не так, то 30 запросов в сутки съедятся очень быстро, так что я бы не говорил что проблема решена...
ps. у меня в ПР уже висит ограничитель, все руки не доходят нормально протестировать.

**Comment by IATkachenko** [comment]
> > Увы, если что-то пойдет не так, то 30 запросов в сутки съедятся очень быстро, так что я бы не говорил что проблема решена...
> 
> Сейчас так работает этот плагин, в официальной же интеграции за лимит запросов не выходит, даже в случае ошибок.

Цепочка вызовов (в обоих интеграциях): [update](https://github.com/yandex/pogoda-home-assistant/blob/main/custom_components/yandex_pogoda/updater.py#L282) -> _get_weather_data -> request. Внутри HA вызов update [обернут в try except](https://github.com/home-assistant/core/blob/6396f54e0dde48aa297b198ab30819d3ed4c9669/homeassistant/helpers/update_coordinator.py#L381) поэтому как только случится request без ответа из за любой ошибки (тут важно только чтобы был отправлен запрос) -- НА не обновит время обновления данных и на следующем круге цикла сделает еще один запрос.

Собственно это issue именно про это -- ловить исключения внутри интеграции и не давать просто так выжирать лимиты ключа.

**Comment by IATkachenko** [comment]
Кажется что сценарий катастрофы успешно в обоих интеграциях работает: при рестарте НА [загрузили старый стейт](https://github.com/yandex/pogoda-home-assistant/blob/main/custom_components/yandex_pogoda/weather.py#L122), поняли что прошло с прошлого обновления слишком много времени, пошли обновляться, крашнулись, HA пошел перезагружать интеграцию.  По крайней мере я не увидел ничего что от этого защищает. Хотя может HА уже научился не рестартить бесконечно интеграцию, если она упала.
